src/checkqueue.h

```diff
@@ -30,103 +30,112 @@ template <typename T>
 class CCheckQueue
 {
 private:
-    //! Mutex to protect the inner state
+    //! Mutex to ensure that sleeping threads are woken.
     boost::mutex mutex;
 
     //! Worker threads block on this when out of work
     boost::condition_variable condWorker;
 
-    //! Master thread blocks on this when out of work
-    boost::condition_variable condMaster;
-
-    //! The queue of elements to be processed.
-    //! As the order of booleans doesn't matter, it is used as a LIFO (stack)
-
-    //! The number of workers (including the master) that are idle.
-    int nIdle;
-
-    //! The total number of workers (including the master).
-    int nTotal;
-
     //! The temporary evaluation result.
-    bool fAllOk;
+    std::atomic<uint8_t> fAllOk;
 
     /**
-     * Number of verifications that haven't completed yet.
-     * This includes elements that are no longer queued, but still in the
-     * worker's own batches.
+     * Number of verification threads that aren't in stand-by. When a thread is
+     * awake it may have a job that will return false, but is yet to report the
+     * result through fAllOk.
      */
-    unsigned int nTodo;
+    std::atomic<unsigned int> nAwake;
+
+    /** If there is presently a master process either in the queue or adding jobs */
+    std::atomic<bool> fMasterPresent;
 
     //! Whether we're shutting down.
     bool fQuit;
 
     //! The maximum number of elements to be processed in one batch
```

Comment:
Looks like, in the current version, you can remove nBatchSize as well (though I'm surprised its not a performance gain to batch operations?